### PR TITLE
Fixed lineout

### PR DESCRIPTION
**Pull request by Vgdn1942**
Fixed lineout and noise on boot instead it https://github.com/boulzordev/android_device_lenovo_A6020/commit/e7efc6a124a7ef8d74060b71540c350a851e1a3c

### CODE REVIEW

File: sound/soc/msm/msm8x16.c [hunk 1]
```diff
@@ -529,6 +507,26 @@ static void msm8x16_ext_spk_delayed_dualmode(struct work_struct *work)
...
+	
+    /* Close the headset device */
+    gpio_direction_output(EXT_SPK_AMP_HEADSET_GPIO, false);
+    usleep_range(EXT_CLASS_D_EN_DELAY,
+        EXT_CLASS_D_EN_DELAY + EXT_CLASS_D_DELAY_DELTA);
+
+    /* Close external audio PA device */
+    for(i = 0; i < AW8155A_MODE; i++) {
+        gpio_direction_output(EXT_SPK_AMP_GPIO, false);
+	gpio_direction_output(EXT_SPK_AMP_GPIO_1, false);
```
Comment (Shr3ps) [inline sound/soc/msm/msm8x16.c:523]: Bad indent

Comment (Vgdn1942) [reply sound/soc/msm/msm8x16.c:523]: What is bad? It works with what is better than a "crutch" in the init.

File: sound/soc/msm/msm8x16.c [hunk 2]
```diff
@@ -542,16 +540,18 @@ static int lineout_status_put(struct snd_kcontrol *kcontrol,
...
 		schedule_delayed_work(&lineout_amp_enable, msecs_to_jiffies(50));
 		break;
-	case 0:
-		//schedule_delayed_work(&lineout_amp_disable, msecs_to_jiffies(5));
-		msm8x16_ext_spk_control(0);
-		break;
 	case 2:
 		schedule_delayed_work(&lineout_amp_dualmode, msecs_to_jiffies(50));
 		break;
+	case 3:
```
Comment (skritchz) [inline sound/soc/msm/msm8x16.c:552]: How do you get the case 3?

Possible values are 0, 1, 2 defined by the index of the string value in https://github.com/Vgdn1942/android_kernel_lenovo_msm8916/blob/cf4430fd40186ad4e83571a1486ab419453ee829/sound/soc/msm/msm8x16.c#L273

File: sound/soc/msm/msm8x16.c (hunk 1 above)
Comment (Shr3ps) [reply sound/soc/msm/msm8x16.c:523]: I mean bad code spaces/tab indentation.

Comment (Vgdn1942) [reply sound/soc/msm/msm8x16.c:523]: I copied with this function https://github.com/boulzordev/android_kernel_lenovo_msm8916/blob/K5_A6020l36_A6020l37/sound/soc/msm/msm8x16.c#L487
Probably he knocked editor.

File: sound/soc/msm/msm8x16.c (hunk 2 above)
Comment (Vgdn1942) [reply sound/soc/msm/msm8x16.c:552]: case 0 - headphones are not connected
case 1 - connected 3-pin headphones without a microphone
case 2 - I do not know
case 3 - connected 4-pin headphones with microphone

I looked here /sys/class/switch/h2w/state
No case 3 were noises when booting from connected 4-pin headphones with microphone

Comment (Vgdn1942) [reply sound/soc/msm/msm8x16.c:552]: The main thing is that it works.

File: sound/soc/msm/msm8x16.c (hunk 1 above)
Comment (Vgdn1942) [reply sound/soc/msm/msm8x16.c:523]: Copy/past from https://github.com/boulzordev/android_kernel_lenovo_msm8916/pull/6/files/cf4430fd40186ad4e83571a1486ab419453ee829#diff-b51eb19739eb22730e773107928095aaL392

File: sound/soc/msm/msm8x16.c (hunk 2 above)
Comment (skritchz) [reply sound/soc/msm/msm8x16.c:552]: Can you just try without the case 3 I'm pretty sure it doesn't do anything.

SOC_ENUM_EXT("Lineout_1 amp", msm_snd_enum[3],
			lineout_status_get, lineout_status_put),
This defines the actions to do when Lineout_1 amp is set by the audio. It calls lineout_status_get or lineout_status_put and possible values are defined in msm_snd_enum[3]. which is 
SOC_ENUM_SINGLE_EXT(3, lineout_text) =>  char *const lineout_text[] = {"DISABLE", "ENABLE", "DUALMODE"};

I was debugging the issue from the kernel side + mixer_path file, and I managed to fix it playing with the Lineout_amp 1 but fix was mush more intrusive. I actually lose all this debugging by a stupid repo sync --force

But the thing is I'm pretty sure that the value can never be 3.

Comment (Vgdn1942) [reply sound/soc/msm/msm8x16.c:552]: I say that without the case 3, when connecting a 4-pin headset with microphone noise obtained at startup. Moreover, no error in the assembly of the kernel does not occur.

Comment (Vgdn1942) [reply sound/soc/msm/msm8x16.c:552]: Why do not you check and see for yourself?

Comment (Vgdn1942) [reply sound/soc/msm/msm8x16.c:552]: Special case 3 removed and the noise was expected when downloading from the connected 4-pins headset. And I do not understand the part about https://github.com/Vgdn1942/android_kernel_lenovo_msm8916/blob/cf4430fd40186ad4e83571a1486ab419453ee829/sound/soc/msm/msm8x16.c#L273 if there is assigned a value of state of ucontrol-> value.integer.value [0 ]; https://github.com/Vgdn1942/android_kernel_lenovo_msm8916/blob/cf4430fd40186ad4e83571a1486ab419453ee829/sound/soc/msm/msm8x16.c#L539 ?

File: sound/soc/msm/msm8x16.c [hunk 3]
```diff
@@ -51,13 +51,12 @@
 #define DEFAULT_MCLK_RATE 9600000
 
 #define EXT_CLASS_D_EN_DELAY 13000
-#define EXT_CLASS_D_DIS_DELAY 3000
```
Comment (danielstuart14) [inline sound/soc/msm/msm8x16.c:54]: Disabling GPIO takes less time than enabling it.
That's why we use different delay times.

Comment (Vgdn1942) [reply sound/soc/msm/msm8x16.c:54]: I had problems with delays in 3000. The code was tested before posting, so the delay and case 3 are needed.

**Comment by Vgdn1942** [comment]
Ебланы.

**Pull request closed without merging**